src/action/list_directory.rs

```diff
@@ -0,0 +1,533 @@
+                path: path,
+                mount_point: None,
+                stream_name: None,
+                file_size_override: None,
+                inode: None,
+                is_virtualroot: None,
+                nested_path: None,
+                ntfs_id: None,
+                ntfs_type: None,
+                offset: None,
+                recursion_depth: None,
+            }),
+            iterator: None,
+        }
+    }
+
+    #[test]
+    fn test_empty_pathspec_field() {
+        let request: Result<super::Request, _> =
+            Request::from_proto(ListDirRequest {
+                pathspec: None,
+                iterator: None,
+            });
+        assert!(request.is_err());
+    }
+
+    #[test]
+    fn test_empty_path_options() {
+        let request: Result<super::Request, _> = Request::from_proto
+            (fill_proto_request(None, Some(0), Some(String::from("/"))));
+        assert!(request.is_ok());
+    }
+
+    #[test]
+    fn test_unset_pathtype() {
+        let request: Result<super::Request, _> = Request::from_proto
+            (fill_proto_request(None, Some(-1), Some(String::from("/"))));
+        assert!(request.is_err());
+    }
+
+
+    #[test]
+    fn test_empty_pathtype() {
+        let request: Result<super::Request, _> = Request::from_proto
+            (fill_proto_request(None, None, Some(String::from("/"))));
+        assert!(request.is_err());
+    }
+
+    #[test]
+    fn test_empty_path() {
+        let request: Result<super::Request, _> = Request::from_proto
+            (fill_proto_request(None, Some(0), None));
+        assert!(&request.is_ok());
+        assert_eq!(request.unwrap().pathspec.path, PathBuf::from("/"));
+    }
+
+    #[test]
+    fn test_empty_dir() {
+        let dir = tempdir().unwrap();
+        let request = super::Request {
+            pathspec: PathSpec {
+                path_options: None,
+                pathtype: PathType::OS,
+                path: PathBuf::from(dir.path()),
+            }
+        };
+        let mut session = session::test::Fake::new();
+        assert!(handle(&mut session, request).is_ok());
+        assert_eq!(session.reply_count(), 0);
+    }
+
+    #[test]
+    fn test_nonexistent_path() {
+        let dir = tempdir().unwrap();
+        let request = super::Request {
+            pathspec: PathSpec {
+                path_options: None,
+                pathtype: PathType::OS,
+                path: PathBuf::from(dir.path().join("nonexistent_subdir")),
+            }
+        };
+        let mut session = session::test::Fake::new();
+        assert!(handle(&mut session, request).is_err());
+    }
+
+    #[test]
+    fn test_lexicographical_order() {
+        let dir = tempdir().unwrap();
+        let dir_path = dir.path();
+        let file_names = vec!["1numfile", "5numfile",
+                              "Afile", "aafile", "afile",
+                              "bfile", "zfile", "*file", "(file)", "=^-^="];
```

Comment:
You can also include a filename with fancy unicode characters.

Reply:
Done.